controlplane/kubeadm/internal/control_plane.go

```diff
@@ -38,10 +38,10 @@ import (
 // It should never need to connect to a service, that responsibility lies outside of this struct.
 // Going forward we should be trying to add more logic to here and reduce the amount of logic in the reconciler.
 type ControlPlane struct {
-	KCP      *controlplanev1.KubeadmControlPlane
-	Cluster  *clusterv1.Cluster
-	Machines FilterableMachineCollection
-
+	KCP               *controlplanev1.KubeadmControlPlane
+	Cluster           *clusterv1.Cluster
+	Machines          FilterableMachineCollection
+	MachineConditions map[string]map[clusterv1.ConditionType]*clusterv1.Condition
```

Comment:
Can we avoid this? We should be able to use `conditions` package to modify the conditions on a Machine directly, and just patch at the end?

Reply:
Did this way to avoid modifying Machines considering that it locks an initial state of machines. But we can use Machines or create a list of Machines out of Machines that is not FilterableMachineCollection and update that. Because we don't use `FilterableMachineCollection` for ReconcileHealth methods and have to change that to a list.
WDYT @fabriziopandini ?

Reply:
As long as we don't modify anything other than conditions, we should be fine

Reply:
I did this change.